examples/controls_example/public/react_controls/control_group/init_controls_manager.ts

```diff
@@ -0,0 +1,158 @@
+        if (!controlState) {
+          subscription.unsubscribe();
+          resolve(undefined);
+        }
+      });
+    });
+  }
+
+  function getControlApi(controlUuid: string) {
+    return children$.value[controlUuid];
+  }
+
+  async function addNewPanel(
+    { panelType, initialState }: PanelPackage<DefaultControlState>,
+    index: number
+  ) {
+    const id = generateId();
+    const nextControlsInOrder = [...controlsInOrder$.value];
+    nextControlsInOrder.splice(index, 0, {
+      id,
+      type: panelType,
+    });
+    controlsInOrder$.next(nextControlsInOrder);
+    controlsPanelState[id] = initialState ?? {};
+    return await untilControlLoaded(id);
+  }
+
+  function removePanel(panelId: string) {
+    delete controlsPanelState[panelId];
+    controlsInOrder$.next(controlsInOrder$.value.filter(({ id }) => id !== panelId));
+    children$.next(omit(children$.value, panelId));
+  }
+
+  return {
+    controlsInOrder$: controlsInOrder$ as PublishingSubject<Array<{ id: string; type: string }>>,
+    getControlApi,
+    setControlApi: (uuid: string, controlApi: DefaultControlApi) => {
+      children$.next({
+        ...children$.getValue(),
+        [uuid]: controlApi,
+      });
+    },
+    serializeControls: () => {
+      const references: Reference[] = [];
+      const explicitInputPanels: {
+        [panelId: string]: ControlPanelState & { explicitInput: object };
+      } = {};
+
+      controlsInOrder$.getValue().forEach(({ id }, index) => {
+        const controlApi = getControlApi(id);
+        if (!controlApi) {
+          return;
+        }
+
+        const {
+          rawState: { grow, width, ...rest },
+          references: controlReferences,
+        } = controlApi.serializeState();
+
+        if (controlReferences && controlReferences.length > 0) {
+          references.push(...controlReferences);
+        }
+
+        explicitInputPanels[id] = {
+          grow,
+          order: index,
+          type: controlApi.type,
+          width,
+          /** Re-add the `explicitInput` layer on serialize so control group saved object retains shape */
+          explicitInput: rest,
+        };
+      });
+
+      return {
+        panelsJSON: JSON.stringify(explicitInputPanels),
+        references,
+      };
+    },
+    api: {
+      getSerializedStateForChild: (childId: string) => {
+        const controlPanelState = controlsPanelState[childId];
+        return controlPanelState ? { rawState: controlPanelState } : undefined;
+      },
+      children$: children$ as PublishingSubject<{
+        [key: string]: DefaultControlApi;
+      }>,
+      getPanelCount: () => {
+        return controlsInOrder$.value.length;
+      },
+      addNewPanel: async (panel: PanelPackage<DefaultControlState>) => {
+        return addNewPanel(panel, controlsInOrder$.value.length);
+      },
+      removePanel,
+      replacePanel: async (panelId, newPanel) => {
```

Comment:
Can't really test `replacePanel` or `addNewPanel` until https://github.com/elastic/kibana/pull/187606 is merged, but the logic LGTM :+1: Especially thanks to the test additions